libraries/AP_Math/fourier.cpp

```diff
@@ -0,0 +1,238 @@
+	
+	_last_result.x=0.0f;
+	_last_result.y=0.0f;
+}
+
+void Fourier_Analysis::Set_Fourier_Analysis(int buffer_size){
+	int i;
+
+	_buffer_size=buffer_size;
+	
+	_signal=(Vector3f *)malloc(buffer_size*sizeof(Vector3f));
+	_timing=(Timing_Struct *)malloc(buffer_size*sizeof(Timing_Struct));
+	_phase=(float *)malloc(buffer_size*sizeof(float));
+	
+	for(i=0;i<_buffer_size;i++)
+	{
+		_signal[i].x=0.0f;
+		_signal[i].y=0.0f;
+		_signal[i].z=0.0f;
+		
+		_timing[i].dt=0.0f;
+		
+		_timing[i].omega.x=0.0f;
+		_timing[i].omega.y=0.0f;
+		_timing[i].omega.z=0.0f;
+		
+		_timing[i].T=0.0f;
+		
+		_phase[i]=0.0f;
+	}
+}
+
+Fourier_Analysis::~Fourier_Analysis(){
+	free(_signal);
+	free(_timing);
+	free(_phase);
+}
+
+void Fourier_Analysis::accumulate(Vector3f *new_sample, Timing_Struct *new_timing){
+	int previous_index;
+	int i;
+	float actual_heading;
+	float actual_yaw;
+	
+	previous_index=_buffer_index-1;
+
+	if(previous_index<0)
+		previous_index=_buffer_size-1;
+	
+	Vector3f signal_current=_signal[_buffer_index];//-_signal_mean/_buffer_size;
+	Vector2f fourier_coef;
+	
+	Vector2f new_result;
+	
+	fourier_coef.x=cosf(_phase[_buffer_index]);
+	fourier_coef.y=-sinf(_phase[_buffer_index]);
+	
+	for(i=0;i<2;i++)
+	{
+		_fourier_transform[i].x -= (fourier_coef.x*signal_current.x);
+		_fourier_transform[i].y -= (fourier_coef.y*signal_current.x);
+	}
+	
+	actual_yaw=get_yaw_angle().x;
+	actual_heading=_phase[_buffer_index]-actual_yaw;
+	
+	_angle_tot.x -= _timing[_buffer_index].dt*(_timing[_buffer_index].omega.x*cosf(actual_heading) - _timing[_buffer_index].omega.y*sinf(actual_heading));
+	_angle_tot.y -= _timing[_buffer_index].dt*(_timing[_buffer_index].omega.y*cosf(actual_heading) + _timing[_buffer_index].omega.x*sinf(actual_heading));
+
+	
+	_signal_mean-=_signal[_buffer_index];
+	
+	_phase[_buffer_index]=_phase[previous_index]+new_timing->omega.z*new_timing->dt;
+	
+	if(_phase[_buffer_index]>M_PI)
+		_phase[_buffer_index]-=2*M_PI;
+	
+	new_timing->T=_timing[previous_index].T+new_timing->dt;
+
+	_signal_mean+=*new_sample;
+	
+	signal_current=*new_sample;//-_signal_mean/_buffer_size;
+	fourier_coef.x=cosf(_phase[_buffer_index]);
+	fourier_coef.y=-sinf(_phase[_buffer_index]);
+	
+	for(i=0;i<2;i++)
+	{
+		_fourier_transform[i].x += (fourier_coef.x*signal_current.x);
+		_fourier_transform[i].y += (fourier_coef.y*signal_current.x);
+	}
+
+	actual_heading=_phase[_buffer_index]-actual_yaw;
+	
+	_angle_tot.x += _timing[_buffer_index].dt*(_timing[_buffer_index].omega.x*cosf(actual_heading) - _timing[_buffer_index].omega.y*sinf(actual_heading));
+	_angle_tot.y += _timing[_buffer_index].dt*(_timing[_buffer_index].omega.y*cosf(actual_heading) + _timing[_buffer_index].omega.x*sinf(actual_heading));
+
+	
+	_signal[_buffer_index]=*new_sample;
+	_timing[_buffer_index]=*new_timing;
+
+	_buffer_index++;
+
+	if(_buffer_index>=_buffer_size)
+		_buffer_index=0;
+	
+	new_result.x=get_pitch_angle();
+	new_result.y=get_yaw_angle().x;
```

Comment:
Get_yaw_angle().x doesn't exist?

Reply:
why that? This function returns a Vector2f value so the .x exists